From 7c1c8c1189634b1c02de8a161203a96c1841db38 Mon Sep 17 00:00:00 2001
Message-Id: <7c1c8c1189634b1c02de8a161203a96c1841db38.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Fri, 13 Jun 2014 09:55:59 +0530
Subject: [PATCH 100/312] FOR_UPSTREAM [VPG]: drm/i915: Maxfifo in update
 watermark.

The following changes have been ported:

(OriginalChangeID - If0362d4f838104c75247dd03692c8696d8727a48)
Enable max_fifo feature when a single plane is enabled. This ensures
that the entire fifo is allocated for the plane that is currently
enabled. Disable max_fifo when multiple planes are enabled.

Issue: GMIN-1462
Change-Id: I8cac7e640ed3f95c7134d3f919aa3c6ffc754a0d
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/intel_pm.c |   16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index fb355e8..7027ffc 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -1523,6 +1523,14 @@ static void valleyview_update_wm(struct drm_crtc *crtc)
 			    &planeb_wm, &cursorb_wm))
 		enabled |= 1 << PIPE_B;
 
+	if (is_maxfifo_needed(dev_priv) & !dev_priv->maxfifo_enabled) {
+		I915_WRITE(FW_BLC_SELF_VLV, FW_CSPWRDWNEN);
+		dev_priv->maxfifo_enabled = true;
+	} else if (dev_priv->maxfifo_enabled && !is_maxfifo_needed(dev_priv)) {
+		I915_WRITE(FW_BLC_SELF_VLV, ~FW_CSPWRDWNEN);
+		dev_priv->maxfifo_enabled = false;
+	}
+
 	I915_WRITE(DSPFW1,
 		   (DSPFW_SR_VAL << DSPFW_SR_SHIFT) |
 		   (DSPFW_CURSORB_VAL << DSPFW_CURSORB_SHIFT) |
@@ -3030,6 +3038,14 @@ static void valleyview_update_sprite_wm(struct drm_plane *plane,
 	enable.cursor_enabled = false;
 	enable.sprite_enabled = enabled;
 
+	if (is_maxfifo_needed(dev_priv) & !dev_priv->maxfifo_enabled) {
+		I915_WRITE(FW_BLC_SELF_VLV, FW_CSPWRDWNEN);
+		dev_priv->maxfifo_enabled = true;
+	} else if (dev_priv->maxfifo_enabled && !is_maxfifo_needed(dev_priv)) {
+		I915_WRITE(FW_BLC_SELF_VLV, ~FW_CSPWRDWNEN);
+		dev_priv->maxfifo_enabled = false;
+	}
+
 	if (intel_plane->plane == 0) {
 		mask = 0x0000ff00;
 		shift = DDL_SPRITEA_SHIFT;
-- 
1.7.9.5

